services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerEndpoint.java

```diff
@@ -0,0 +1,388 @@
+import org.keycloak.protocol.oid4vc.model.ErrorType;
+import org.keycloak.protocol.oid4vc.model.Format;
+import org.keycloak.protocol.oid4vc.model.OID4VCClient;
+import org.keycloak.protocol.oid4vc.model.PreAuthorizedCode;
+import org.keycloak.protocol.oid4vc.model.PreAuthorizedGrant;
+import org.keycloak.protocol.oid4vc.model.SupportedCredentialConfiguration;
+import org.keycloak.protocol.oid4vc.model.VerifiableCredential;
+import org.keycloak.protocol.oidc.grants.PreAuthorizedCodeGrantType;
+import org.keycloak.provider.ProviderFactory;
+import org.keycloak.services.managers.AppAuthManager;
+import org.keycloak.services.managers.AuthenticationManager;
+import org.keycloak.utils.MediaType;
+
+import java.net.URI;
+import java.time.Instant;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ * Provides the (REST-)endpoints required for the OID4VCI protocol.
+ * <p>
+ * {@see https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0.html}
+ *
+ * @author <a href="https://github.com/wistefan">Stefan Wiedemann</a>
+ */
+public class OID4VCIssuerEndpoint {
+
+    private static final Logger LOGGER = Logger.getLogger(OID4VCIssuerEndpoint.class);
+
+    public static final String CREDENTIAL_PATH = "credential";
+    public static final String CREDENTIAL_OFFER_PATH = "credential-offer/";
+    private final KeycloakSession session;
+    private final AppAuthManager.BearerTokenAuthenticator bearerTokenAuthenticator;
+    private final ObjectMapper objectMapper;
+    private final TimeProvider timeProvider;
+
+    private final String issuerDid;
+    // lifespan of the preAuthorizedCodes in seconds
+    private final int preAuthorizedCodeLifeSpan;
+
+    private final Map<Format, VerifiableCredentialsSigningService> signingServices;
+
+    public OID4VCIssuerEndpoint(KeycloakSession session,
+                                String issuerDid,
+                                Map<Format, VerifiableCredentialsSigningService> signingServices,
+                                AppAuthManager.BearerTokenAuthenticator authenticator,
+                                ObjectMapper objectMapper, TimeProvider timeProvider, int preAuthorizedCodeLifeSpan) {
+        this.session = session;
+        this.bearerTokenAuthenticator = authenticator;
+        this.objectMapper = objectMapper;
+        this.timeProvider = timeProvider;
+        this.issuerDid = issuerDid;
+        this.signingServices = signingServices;
+        this.preAuthorizedCodeLifeSpan = preAuthorizedCodeLifeSpan;
+
+    }
+
+    /**
+     * Provides the URI to the OID4VCI compliant credentials offer
+     */
+    @GET
+    @Produces(MediaType.APPLICATION_JSON)
+    @Path("credential-offer-uri")
+    public Response getCredentialOfferURI(@QueryParam("credential_configuration_id") String vcId) {
```

Comment:
The specification of oid4vci dose not specify credential offer endpoint's interface.
Therefore, is it keycloak's specific interface ?

Reply:
Yes. In order to create a qr, we need a way to get an offer-uri. Thus I added that endpoint.

Reply:
Thank you for the clarification.